Global dry air mass conservation

yikwill opened the pull request:
# PhysicsNeMo Pull Request

## Description

Add a hard `DryAirMassConstraint` and unify HEALPix constraint call sites so constraints can use the model input.

- New `DryAirMassConstraint`: at each step apply a globally uniform surface-pressure fixer so column dry-air mass is conserved (ACE2-style)
- All hard constraints now take `(prediction, input)` with shape `[B, F, T, C, H, W]` (`input` unused by some constraints, required for a consistent API)
- Refactor `NonnegativeConstraint` for that API / torch.compile (named `in_channels`/`out_channels`, buffer thresholds, RankZero logging)
- Optional `keep_grad_through_clamp=False` straight-through estimator on the nonnegative clamp: forward stays ≥ physical zero, backward passes identity grads so below-threshold cells still get a learning signal
- Add constraint unit tests (including STE forward/grad checks)

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.

yikwill committed:
First attempt at dry air mass conservation block

yikwill committed:
Adds non-negative constraint to dry air mass conservation module

yikwill committed:
Changes non-negative correction to be in physical units

yikwill committed:
Updates variable names and ensures constraint module outputs correct sp

yikwill committed:
Bug fixes for dry air mass constraint

yikwill committed:
Variable name changes and forces float32

yikwill committed:
Adds check and changes variable names

yikwill committed:
Adds comments and cleans up implementation

yikwill committed:
Adds checks and logging to constraints

yikwill committed:
Updates non-recurrent constraint code to match recurrent model

yikwill committed:
Removes support for box-cox transformed sp

yikwill committed:
Fixing var name bug

daviddpruitt commented on physicsnemo/models/dlwp_healpix_layers/healpix_constraints.py:
```diff
@@ -40,12 +55,90 @@ def __init__(
+        '''
+        Tensors are expected to be in the shape [B, F, T, C, H, W]
+        '''
+        with torch.cuda.amp.autocast(enabled=False):
```
Disabling lower precision is a bit heavyweight, is this section sensitive to precision? If so we might be able to work around it.

yikwill replied:
I'm not actually sure if the constraint is sensitive to precision, but I will test it. I was motivated to force fp32 because @akshaysubr did it for the hydrostatic loss since it requires unnormalizing to physical units (which this constraint also does). I'll ask him about it.

yikwill replied:
I've tested that fp16 and bf16 will indeed break this constraint because it computes small differences of large surface pressures (in Pa). These small differences will not be correctly represented with lower precision than fp32.

daviddpruitt commented on physicsnemo/models/dlwp_healpix/HEALPixUNet.py:
```diff
@@ -380,21 +380,20 @@ def forward(self, inputs: Sequence, output_only_last=False, conditions_cln: Sequ
             if self.constraints is not None:
                 for constraint in self.constraints:
-                    reshaped = constraint(reshaped)
+                    prediction = constraint(prediction, orig_input)
```
This will end up breaking constraints that only look at the prediction, I like the approach of making all of them take the prediction and the input and ignoring the input if they don't need it. We'll have to figure out a consistent interface for the constraints so they can handle a variable set of inputs and then update them all. For now I think its fine.

yikwill replied:
I agree. I changed the non-negative constraint to also take both `prediction` and `orig_input` even though it doesn't use `orig_input`. By variable sets of inputs do you mean updating multiple time steps at once? Perhaps in the future we could have every constraint take the most recent output (`prediction`) as well as a list of all previous outputs (`outputs`)?

yikwill committed:
Adds comment explaining autocast disable

yikwill committed:
Merge branch 'dev' into wy/dry_air_mass_constraint

yikwill committed:
Refactor nonnegative constraint to be torch compile compatible

yikwill committed:
Adds nonnegative constraint tests

yikwill committed:
Refactors dry air mass constraint to be compatible with torch compile

yikwill committed:
Makes dry air mass constraint torch compile compatible and adds tests for it

yikwill committed:
Merge branch 'dev' into wy/dry_air_mass_constraint

yikwill committed:
Ensures constraints take in both prediction and input tensors from UNets

yikwill committed:
Make logging only on rank 0

yikwill committed:
Use distributed logging wrapper

yikwill committed:
Merge branch 'dev' into wy/dry_air_mass_constraint

yikwill committed:
Sync healpix_constraints logging with dlesym-aq.

Use PythonLogger plus RankZeroLoggingWrapper as logger0 so warnings
only emit on rank 0 when DistributedManager is initialized.

yikwill committed:
Add opt-in STE for NonnegativeConstraint clamps.

keep_grad_through_clamp (default False) keeps forward values
nonnegative while passing identity gradients through saturated cells.

yikwill committed:
Merge branch 'dev' into wy/dry_air_mass_constraint

Resolve HEALPixUNet: keep ml_diagnosed prognostic/diagnostic split with
two-arg constraint(out, orig_input).

yikwill committed:
Merge branch 'dev' into wy/dry_air_mass_constraint

yikwill committed:
Merge branch 'dev' into wy/dry_air_mass_constraint

yikwill committed:
Skip nonnegative constraint warning when all variables are present.

Only log missing-channel warnings when at least one requested variable is absent from model output channels.
